scripts/genererate_drivers.py

```diff
@@ -0,0 +1,633 @@
+            }}
+        """
+
+        rv.append(textwrap.dedent(read_func))
+
+    return '\n'.join(rv)
+
+
+def generate_metal_function(device: str, reg_list: t.List[Register]) -> str:
+    """
+    Generates the exported register access functions for
+    a given device and register list.
+
+    :param device: the name of the device
+    :param reg_list: the list of registers for the device.
+    :return:  the c code for the exported register access functions
+    """
+
+    rv: t.List[str] = []
+
+    for a_reg in reg_list:
+        name = a_reg.name.lower()
+        size = a_reg.width
+
+        write_func = f"""
+            void metal_{device}_{name}_write(const struct metal_{device} *{device}, uint{size}_t data)
+            {{
+                if ({device} != NULL)
+                    {device}->vtable.v_{device}_{name}_write({device}->{device}_base, data);
+            }}
+            """
+        rv.append(textwrap.dedent(write_func))
+
+        read_func = f"""
+            uint{size}_t metal_{device}_{name}_read(const struct metal_{device} *{device})
+            {{
+                if ({device} != NULL)
+                    return {device}->vtable.v_{device}_{name}_read({device}->{device}_base);
+                return (uint{size}_t)-1;
+            }}
+            """
+
+        rv.append(textwrap.dedent(read_func))
+
+    return '\n'.join(rv)
+
+
+def generate_metal_dev_drv(vendor, device, index, reglist):
+    """
+    Generate the driver source file contents for a given device
+    and register list
+
+    :param vendor: the vendor creating the device
+    :param device: the device
+    :param index: the index of the device used
+    :param reglist: the list of registers
+    :return: a string containing of the c code for the basic driver
+    """
+    template = string.Template(textwrap.dedent(METAL_DEV_DRV_TMPL))
+
+    return template.substitute(
+        vendor=vendor,
+        device=device,
+        cap_device=device.upper(),
+        index=str(index),
+        offsets=generate_offsets(device, reglist),
+        base_functions=generate_base_functions(device, reglist),
+        metal_functions=generate_metal_function(device, reglist),
+        def_vtable=generate_def_vtable(device, reglist)
+    )
+
+
+###
+# main
+###
+
+
+def handle_args():
+    """
+    :return:
+    """
+    parser = argparse.ArgumentParser()
+
+    parser.add_argument(
+        "-o",
+        "--object-model",
+        type=argparse.FileType('r'),
+        help="",
+    )
+
+    parser.add_argument(
+        "-d",
+        "--duh-file",
+        type=argparse.FileType('r'),
+        help="",
+    )
+
+    parser.add_argument(
+        "-v",
+        "--vendor",
+        help="",
+    )
+
+    parser.add_argument(
+        "-D",
+        "--device",
+        help="",
+    )
+
+    parser.add_argument(
+        "-m",
+        "--metal-dir",
+        help="",
+    )
+
+    parser.add_argument(
+        "-x",
+        "--check-existence",
```

Comment:
This isn't a clear argument name. What is this checking the "existence" of? And what is the consequence of checking the existence of it?

I think a better name would be `--overwrite-existing`, which makes it clearer that 1) this is about the existence of the drivers, since "overwrite" ties this flag to the primary action of the script itself and 2) the behavior is that if the drivers already exist, then if this flag is true, the drivers will be overwritten.